keel-core/src/main/kotlin/com/netflix/spinnaker/keel/caffeine/CacheFactory.kt

```diff
@@ -66,16 +88,50 @@ class CacheFactory(
         .recordStats()
     }
 
-  private fun <K, V, C: AsyncCache<K, V>> C.monitor(cacheName: String) =
+  private fun <K, V, C : AsyncCache<K, V>> C.monitor(cacheName: String) =
     CaffeineCacheMetrics.monitor(meterRegistry, this, cacheName)
 }
 
-fun <K, V> (suspend (K) -> V?).toAsyncCacheLoader() : AsyncCacheLoader<K, V> =
+fun <K : Any, V> (suspend (K) -> V?).toAsyncCacheLoader(): AsyncCacheLoader<K, V> =
   AsyncCacheLoader<K, V> { key, executor ->
     CoroutineScope(executor.asCoroutineDispatcher())
       .future { this@toAsyncCacheLoader.invoke(key) }
   }
 
+fun <K : Any, V> (suspend () -> Map<K, V>).toAsyncBulkCacheLoader(): AsyncCacheLoader<K, V> =
+  object : AsyncCacheLoader<K, V> {
+    override fun asyncLoad(key: K, executor: Executor): CompletableFuture<V> {
+      throw UnsupportedOperationException()
+    }
+
+    override fun asyncLoadAll(
+      keys: Iterable<K>,
+      executor: Executor
+    ): CompletableFuture<Map<K, V>> =
+      CoroutineScope(executor.asCoroutineDispatcher())
+        .future { this@toAsyncBulkCacheLoader.invoke() }
```

Comment:
This is throwing me out for a loop... It looks reentrant, but I know you know what you're doing. TBH I can't say I immediately understand most of the kotlinism-heavy code in this PR. 😓

Reply:
This is an extension method on a function with a particular signature that wraps it up in a coroutine and returns the function's result in a `Future`. `this@toAsynBulkCacheLoader` is a reference to the function itself. It's not re-entrant, just decorates the function.

Reply:
![](https://media.giphy.com/media/LpLd2NGvpaiys/giphy.gif)